flake-module.nix

```diff
@@ -153,37 +169,42 @@ in
             (projectKey: cfg:
               let
                 inherit (pkgs.lib.lists) optionals;
+                projectOverlay =
+                  pkgs.lib.composeExtensions (pkgs.haskell.lib.packageSourceOverrides cfg.source-overrides) cfg.overrides;
                 # Apply user provided source-overrides and overrides to
                 # `cfg.haskellPackages`.
-                hp = cfg.haskellPackages.extend
-                  (pkgs.lib.composeExtensions
-                    (pkgs.haskell.lib.packageSourceOverrides cfg.source-overrides)
-                    cfg.overrides);
+                projectPackages = cfg.haskellPackages.extend projectOverlay;
+                localPackagesOverlay = self: _:
+                  lib.mapAttrs
+                    (name: value: value.modifier (self.callCabal2nix name value.root { }))
+                    cfg.packages;
+                finalPackages = projectPackages.extend localPackagesOverlay;
```

Comment:
I have some package overrides that need to access the package set, so they can't be in `packages.<name>.modifier`, but `overrides` doesn't work either because that is applied before the packages are added.

Perhaps the order of the overlays could be reversed, so that `overrides` applies last?

I recommend to use `composeExtensions` (or `composeManyExtensions`) because multiple uses of `extend` each create a very large attribute set that goes to waste.

Reply:
> I have some package overrides that need to access the package set, so they can't be in `packages.<name>.modifier`, but `overrides` doesn't work either because that is applied before the packages are added.

Do you have an example use-case?

Reply:
- `buildFromSdist` (though that would be superseded by #35)
 - optparse applicative completions [since recently](https://github.com/NixOS/nixpkgs/commit/ac1f1ad0e0a8cfd35db476529c82354e033a48cc#diff-f690c88564d6774086ecabe92c7e4c18078889589f8978ad0819f57bf515eaf9)
 - `addBuildToolDepends`
 - `overrideScope` or `.override` to change a dependency for just one package

It's not much of an example, but my real life use case is https://github.com/hercules-ci/hercules-ci-agent/blob/master/nix/packages.nix#L82.

Reply:
I think I understand the problem now.

> Perhaps the order of the overlays could be reversed, so that `overrides` applies last?

Yea, this makes more sense to me, because `overrides` becomes _the_ place where you get "the final say" in overriding things. One could override even the local packages (not just dependencies) in `overrides`. With this change, will there be any value to keeping `packages.${name}.modifier`? Or should we remove the `modifier` option?

Reply:
> will there be any value to keeping `packages.${name}.modifier`? Or should we remove the `modifier` option?

Let's remove `modifier`. Having a less powerful way of doing the same thing is just confusing.

It being in `packages.${name}` was nice, but that's not enough reason to keep it.

Reply:
> should I achieve equivalent functionality with overrides?

That's how it should be done, yes.

Reply:
> Perhaps the order of the overlays could be reversed, so that `overrides` applies last?

Did in 13a2646304e654c603e95620d67ec7f9e9ae8ac7 ... hope this is the correct order:

```nix
                  pkgs.lib.composeManyExtensions 
                    [ # The order here matters.
                      #
                      # User's overrides (cfg.overrides) is applied **last** so
                      # as to give them maximum control over the final package
                      # set used.
                      cfg.overrides
                      (pkgs.haskell.lib.packageSourceOverrides cfg.source-overrides) 
                      localPackagesOverlay
                    ];
```